lib/doctest.js

```diff
@@ -523,50 +562,113 @@ function commonjsEval(source, path) {
   } finally {
     fs.unlinkSync (abspath);
   }
-  return run (queue);
+  return run (options, queue, options);
 }
 
-function run(queue) {
-  return queue.reduce (function(accum, io) {
-    var thunk = accum.thunk;
-    if (io.type === INPUT) {
-      if (thunk != null) thunk ();
-      accum.thunk = io.thunk;
-    } else if (io.type === OUTPUT) {
-      var either;
+function run(options, queue, optionalLogMediator) {
+  var logMediator = optionalLogMediator || {};
+  return queue.reduce (function(p, io) {
+    return p.then (function(accum) {
+      var thunk = accum.thunk;
+
+      if (io.type === INPUT) {
+        if (thunk != null) thunk ();
+        accum.thunk = io.thunk;
+        return accum;
+      }
+
+      var errored;
+      var expecteds = io.thunk ();
+      var outputs = [];
+
+      accum.thunk = null;
+
+      logMediator.emit = function(channel, value) {
+        outputs.push ({channel: channel, value: value});
+      };
+
       try {
-        either = {tag: 'Right', value: thunk ()};
+        outputs.push ({channel: null, value: thunk ()});
+        errored = false;
       } catch (err) {
-        either = {tag: 'Left', value: err};
+        outputs.push ({channel: null, value: err});
+        errored = true;
       }
-      accum.thunk = null;
-      var expected = io.thunk ();
-
-      var pass, repr;
-      if (either.tag === 'Left') {
-        var name = either.value.name;
-        var message = either.value.message;
-        pass = io['!'] &&
-               name === expected.name &&
-               message === expected.message.replace (/^$/, message);
-        repr = '! ' + name +
-               (expected.message && message.replace (/^(?!$)/, ': '));
-      } else {
-        pass = !io['!'] && Z.equals (either.value, expected);
-        repr = show (either.value);
+
+      function awaitOutput() {
+        return Promise (function(res) {
+          var t = setTimeout (done, options.logTimeout);
+          logMediator.emit = function(channel, value) {
+            outputs.push ({channel: channel, value: value});
+            clearTimeout (t);
+            t = setTimeout (done, options.logTimeout);
+          };
+          function done() {
+            logMediator.emit = function() {};
+            clearTimeout (t);
+            res ();
+          }
+        });
       }
 
-      accum.results.push ([
-        pass,
-        repr,
-        io['!'] ?
-          '! ' + expected.name + expected.message.replace (/^(?!$)/, ': ') :
-          show (expected),
-        io[':']
-      ]);
-    }
-    return accum;
-  }, {results: [], thunk: null}).results;
+      return typeof options.logFunction === 'string' ?
+             (awaitOutput ()).then (verifyOutput) :
+             Promise.resolve (verifyOutput ());
```

Comment:
Does the logging functionality need to be optional? Are you worried about naming collisions?

Reply:
Yes. The user explicitly injects a log function with a given name, which must not collide with other values defined in their module.

Reply:
You want to save users from resorting to `--log-function asdf`, I assume. :joy: